ArduPlane/soaring.cpp

```diff
@@ -66,156 +66,118 @@ void Plane::update_soaring() {
-                    break;
-                case SoaringController::LoiterStatus::DRIFT_EXCEEDED:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooFar, mode_fbwb.name());
-                    set_mode(mode_fbwb, ModeReason::SOARING_DRIFT_EXCEEDED);
-                    break;
-                } // switch louterStatus
-            break;
-            }
+        // Heading lined up and loiter status not good to continue. Need to switch mode.
 
-        case Mode::Number::CRUISE: {
-            const bool headingLinedupToCruise = plane.mode_loiter.isHeadingLinedUp_cd(cruise_state.locked_heading_cd);
-            if (cruise_state.locked_heading && !headingLinedupToCruise && loiterStatus!=SoaringController::LoiterStatus::ALT_TOO_LOW && timer<20e3) {
-                break;
-            }
-            // return to cruise with old ground course
-            const CruiseState cruise = cruise_state;
-            switch (loiterStatus) {
-                case SoaringController::LoiterStatus::ALT_TOO_HIGH:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooHigh, mode_cruise.name());
-                    set_mode(mode_cruise, ModeReason::SOARING_ALT_TOO_HIGH);
-                    break;
-                case SoaringController::LoiterStatus::ALT_TOO_LOW:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooLow, mode_rtl.name());
-                    set_mode(mode_rtl, ModeReason::SOARING_ALT_TOO_LOW);
-                    break;
-                default:
-                case SoaringController::LoiterStatus::THERMAL_WEAK:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooWeak, mode_cruise.name());
-                    set_mode(mode_cruise, ModeReason::SOARING_THERMAL_ESTIMATE_DETERIORATED);
-                    break;
-                case SoaringController::LoiterStatus::DRIFT_EXCEEDED:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooFar, mode_cruise.name());
-                    set_mode(mode_cruise, ModeReason::SOARING_DRIFT_EXCEEDED);
-                    break;
-            } // switch loiterStatus
-            cruise_state = cruise;
-            set_target_altitude_current();
-            break;
-            } // case Cruise
+        // Determine appropriate mode.
+        Mode* exit_mode = previous_mode;
 
-        case Mode::Number::AUTO: {
-            //Get the lat/lon of next Nav waypoint after this one:
-            AP_Mission::Mission_Command current_nav_cmd = mission.get_current_nav_cmd();;
-            const bool headingLinedupToWP = plane.mode_loiter.isHeadingLinedUp(next_WP_loc, current_nav_cmd.content.location);
-            if (!headingLinedupToWP && loiterStatus!=SoaringController::LoiterStatus::ALT_TOO_LOW && timer<20e3) {
+        if (loiterStatus == SoaringController::LoiterStatus::ALT_TOO_LOW && 
+           ((previous_mode == &mode_cruise) || (previous_mode == &mode_fbwb))) {
```

Comment:
Nomination of specific modes like this can lead to issues where somebody adds another throttle-controlled mode but forgets to update this sort of list.  That's why we try to have callbacks on the mode saying they have a particular property.

This isn't a blocker, just an observation.

Reply:
This is allllll over the place. Part of the onion v2 stuff will take care of all that so lets not worry about it for now